pkg/operation/botanist/resource_manager.go

```diff
@@ -66,3 +68,65 @@ func (b *Botanist) WaitUntilManagedResourcesDeleted(ctx context.Context) error {
 		return retry.MinorError(fmt.Errorf("not all managed resources have been deleted in the shoot cluster (still existing: %s)", names))
 	})
 }
+
+// WaitUntilAllManagedResourcesDeleted waits until all managed resources are gone or the context is cancelled.
+func (b *Botanist) WaitUntilAllManagedResourcesDeleted(ctx context.Context) error {
+	return b.waitUntilManagedResourceAreDeleted(ctx, client.InNamespace(b.Shoot.SeedNamespace))
+}
+func (b *Botanist) waitUntilManagedResourceAreDeleted(ctx context.Context, listOpt ...client.ListOption) error {
```

Comment:
I am deleting all of the managed resources (not only those labeled with `origin=gardener` ), because later in the flow during the PrepareControlPlaneDeployments... the resource manager is scaled down. And this will block the termination of the namespace because there will be managed resources with finalizers left which don't have the label. I am not sure why during the deletion flow we delete only those with origin=gardener, maybe @rfranzke can explain.

Reply:
I'm not sure what @stoyanr means with code duplication - isn't the `WaitUntilManagedResourcesDeleted` just calling `waitUntilManagedResourceAreDeleted` with proper parameters? Maybe this comment was made at an earlier point in time.
Anyways, generally, there are `ManagedResource`s created by Gardener (labeled with `origin=gardener`) and other `ManagedResource`s created by extensions (unknown to Gardener, not labeled).

> Is this different behavior really needed? Why is the label important for the deletion flow, but not for the migration flow?

As part of the shoot deletion we are deleting the extension resources in the seed regularly which makes them cleaning up their created `ManagedResource` objects. Gardener is only responsible for deleting those that it created, so it deletes only those with the `origin=gardener` label.
As part of the shoot migration preparation the extensions don't delete their `ManagedResource`s. Instead, Gardener sets the `.spec.keepObjects=true` and then deletes all of them. This will make the GRM to not remove any managed resource from the shoot, instead, it'll only remove its finalizer from the `ManagedResource` and delete them.

However, I get your point - it would be more consistent/correct if the extension would set the `.spec.keepObject=true` + delete the `ManagedResource`s they created earlier themselves as part of their migration phase. Then we don't need this label separation code here. Though, I'm okay with proceeding as it is for now (in the interest of simplicity). If needed we can refactor it as a follow-up?

> I would also recommend using method names and comments that make it more explicit where the difference comes from, right now WaitUntilManagedResourcesDeleted is almost the same as WaitUntilAllManagedResourcesDeleted, the comments are identical, and so the difference between them rather obscure.

Agreed, `WaitUntilManagedResourcesDeleted` could be something like `WaitUntilGardenersManagedResourcesDeleted` or so. Definitely, the naming could be improved for readability.

Reply:
@rfranzke Thanks for your explanations. Yes, the remark regarding code duplication referred to an earlier state of the code, it was subsequently refactored to remove the duplication.